include/deal.II/base/ndarray.h

```diff
@@ -0,0 +1,102 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1921 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_ndarray_h
+#define dealii_ndarray_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+
+#include <array>
+
+DEAL_II_NAMESPACE_OPEN
+
+#ifndef DOXYGEN
+namespace
+{
+  // clang-format off
+  //
+  // A variadic template helper class to recursively "unroll" the size
+  // information of the ndarray. This is best explained on an example:
+  //
+  //    HelperArray<double, 1, 2, 3, 4>::type
+  // == std::array<HelperArray<double, 2, 3, 4>::type, 1>
+  // == std::array<std::array<HelperArray<double, 3, 4>::type, 2>, 1>
+  // == std::array<std::array<std::array<HelperArray<double, 4>::type, 3>, 2>, 1>
+  // == std::array<std::array<std::array<std::array<HelperArray<double>::type, 4>, 3>, 2>, 1>
+  // == std::array<std::array<std::array<std::array<double, 4>, 3>, 2>, 1>
+  //
+  // clang-format on
+  template <typename T, std::size_t... Ns>
+  struct HelperArray
+  {};
```

Comment:
I don't think you even need the class body here. A forward declaration would be sufficient and probably lead to clearer error messages.